stlab/channel.hpp

```diff
@@ -166,141 +200,400 @@ struct shared_process_sender {
+    }
+};
+
+/**************************************************************************************************/
+
+template <typename... T>
+struct zip_queue_strategy
+{
+    static const std::size_t Size = sizeof...(T);
+    using value_type    = std::tuple<first_t<T...>>;
+    using queue_size_t  = std::array<std::size_t, Size>;
+    using queue_t       = std::tuple<std::deque<T>...>;
+    std::size_t         _index{ 0 };
+    std::size_t         _popped_index{ 0 };
+    queue_t             _queue;
+
+
+    bool empty() const { return get_i<0, Size>::go(_queue, _index, [](const auto& c) { return c.empty(); }, true); }
+
+    auto front() {
+        assert(!empty() && "front on an empty container is a very bad idea!");
+        return std::make_tuple(get_i<0, Size>::go(_queue, 
+                                                  _index, 
+                                                  [](auto& c) { return c.front(); }, 
+                                                  first_t<T...>{} ));
+    }
+
+    void pop_front() {
+        void_i<0, Size>::go(_queue, _index, [](auto& c) { c.pop_front(); });
+        _popped_index = _index;
+        ++_index;
+        if (_index == Size) _index = 0; // restart from the first sender
+    }
+
+    auto size() const {
+        queue_size_t result;
+        std::size_t i = 0;
+        tuple_for_each(_queue, [&i, &result, this](const auto& c)
+        {
+            if (i == _popped_index) result[i] = c.size();
+            else result[i] = std::numeric_limits<std::size_t>::max();
+            ++i;
+        });
+        return result;
+    }
+
+    template <std::size_t I, typename U>
+    void append(U&& u) {
+        std::get<I>(_queue).emplace_back(std::forward<U>(u));
+    }
+};
+
+/**************************************************************************************************/
+
+template <typename... T>
+struct merge_queue_strategy
+{
+    static const std::size_t Size = sizeof...(T);
+    using value_type = std::tuple<first_t<T...>>;
+    using queue_size_t = std::array<std::size_t, Size>;
+    using queue_t = std::tuple<std::deque<T>...>;
+    std::size_t         _index{ 0 };
+    std::size_t         _popped_index{ 0 };
+    queue_t             _queue;
+
+
+    bool empty() const { return tuple_find(_queue, [](const auto& c) { return !c.empty(); } ) == Size; }
+
+    auto front() {
+        assert(!empty() && "front on an empty container is a very bad idea!");
+        _index = tuple_find(_queue, [](const auto& c) { return !c.empty(); });
+        return std::make_tuple(get_i<0, Size>::go(_queue, 
+                                                  _index, [](auto& c) { return c.front(); }, 
+                                                  first_t<T...>{} ));
+    }
+
+    void pop_front() {
+        void_i<0, Size>::go(_queue, _index, [](auto& c) { c.pop_front(); });
+        _popped_index = _index;
+        ++_index;
+        if (_index == Size) _index = 0; // restart from the first sender
```

Comment:
In the meantime I realized that the ++_index and the following if () does not make any sense here. It is left over from a copy paste. Will remove it soon.